src/volume_mgr.rs

```diff
@@ -426,176 +500,187 @@ where
                 if dir_entry.is_some() {
                     return Err(Error::FileAlreadyExists);
                 }
-                let file_name =
-                    ShortFileName::create_from_str(name).map_err(Error::FilenameError)?;
                 let att = Attributes::create_from_fat(0);
-                let entry = match &mut volume.volume_type {
-                    VolumeType::Fat(fat) => {
-                        fat.write_new_directory_entry(self, dir, file_name, att)?
-                    }
+                let volume_idx = self.get_volume_by_id(volume_id)?;
+                let entry = match &mut self.open_volumes[volume_idx].volume_type {
+                    VolumeType::Fat(fat) => fat.write_new_directory_entry(
+                        &self.block_device,
+                        &self.time_source,
+                        directory_info,
+                        sfn,
+                        att,
+                    )?,
                 };
 
-                let search_id = self.id_generator.get();
+                let file_id = File(self.id_generator.get());
 
-                let file = File {
-                    starting_cluster: entry.cluster,
+                let file = FileInfo {
+                    file_id,
+                    volume_id,
                     current_cluster: (0, entry.cluster),
                     current_offset: 0,
-                    length: entry.size,
                     mode,
                     entry,
-                    search_id,
+                    dirty: false,
                 };
 
                 // Remember this open file
-                self.open_files
-                    .push(ClusterDescriptor::new(
-                        volume.idx,
-                        file.starting_cluster,
-                        search_id,
-                    ))
-                    .map_err(|_| Error::TooManyOpenFiles)?;
+                unsafe {
+                    self.open_files.push_unchecked(file);
```

Comment:
Maybe make a note that we checked above, so the unsafe is fine.

Reply:
done